### PR TITLE
Add a new @PendingFeatureIf annotation / extension to allow features to be conditionally pending

### PR DESCRIPTION
**Pull request by ryangardner**
In our specific case we have some geb-based tests that target different deployed environments and occasionally it is convenient to have tests marked as being conditionally Pending. 

I could see other places this might be useful - (features that are implemented on one OS but not another, etc)

---
This change is [Reviewable](https://reviewable.io/reviews/spockframework/spock/988)

### CODE REVIEW

**Comment by leonard84** [comment]
Hi, thanks for your contribution. We are currently working on spock-2.0, until that is merged to master we won't merge PRs.

**Comment by leonard84** [comment]
Hm, why don't you want to use `@IgnoreIf`/`@Requires` for this use-case?

**Comment by ryangardner** [comment]
The behavior of `PendingFeature` - where it expects the test to fail (because the feature isn't implemented) and treats a failed test as passing is different from `Ignore` - where it doesn't run the test - we actually _do_ want to run the test and be notified if it suddenly passes. 

The scenario where we had conditional pending features is for a UI automation suite that hits multiple environments. (could be more than two, but for simplicity sake, lets assume a dev and prod environment)

A new feature is deployed but only active on the dev environment. We have some UI tests that actively run against that new feature - if we mark them as `IgnoreIf(environment=prod)` - the test suite will pass - with the tests running against our dev environment and not running in prod. 

Some time passes - and the feature is released to production. The test suite still passes - with the tests running when we execute it against our dev environment but the tests are still ignored when running against the production environment. If some developer checks in code that breaks that feature - but only breaks it in prod - our regression suite wont detect it because the test is ignored.

In a perfect world, we'd remember to remove that `@IgnoreIf` annotation promptly after the feature was released. What's nice about the `@PendingFeature` annotation (and in this PR, the `@PendingFeatureIf` annotation) is that it _gives you a reminder_ that this thing you were ignoring previously is now passing and you should update your tests to reflect that. It's a pretty harsh reminder (it fails your tests / build) - but it's an effective one.

The same thing that might motivate you to sometimes use `@PendingFeature` and to sometimes use `@Ignore` exist when you need to do those conditional activations - there is the `@IgnoreIf` which is great when you want to conditionally ignore a test, but the lack of a `@PendingFeatureIf` makes it hard if you want to conditionally mark a feature as pending. 

If there's a way to combine `@IgnoreIf` and `@Requires` together and get the same behavior as conditionally applying a `@PendingFeature` annotation, we could use that instead

File: docs/extensions.adoc [hunk 1]
```diff
@@ -155,6 +155,29 @@ Groovy has the `groovy.transform.NotYetImplemented` annotation which is similar
...
+
+[source,groovy]
+----
+@PendingFeatureIf({ os.windows })
+def "I'm not yet implemented on windows, but I am on other operating systems"() { ... }
+----
+
+[source,groovy]
+----
+@PendingFeatureIf({ System.getProperty("targetEnvironment").startsWith("production") })
```
Comment (leonard84) [inline docs/extensions.adoc]: Take a look at `PreconditionContext` we have a shorthand for that `sys`, and I would just use `== 'prod'`

And just merge those two blocks

File: spock-core/src/main/java/spock/lang/PendingFeatureIf.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+import org.spockframework.runtime.extension.builtin.PendingFeatureIfExtension;
+import org.spockframework.util.Beta;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * If a passed in closure evaluates to true, this will be treated as a {@link PendingFeature}
```
Comment (leonard84) [inline spock-core/src/main/java/spock/lang/PendingFeatureIf.java]: Use similar wording to `IgnoreIf` -> `... if the given condition holds.`

File: spock-core/src/main/java/spock/lang/PendingFeatureIf.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+ *
+ * Otherwise it will be treated like a normal feature.
+ *
+ * The closure follows the same rules as those used by {@link IgnoreIf}
```
Comment (leonard84) [inline spock-core/src/main/java/spock/lang/PendingFeatureIf.java]: I would just copy the block with the information regarding the `PreconditionContext` from `@Requires`

**Comment by leonard84** [comment]
@ryangardner do you have time to fix the comments soon, I'd like to use the annotation in #1075, but if you need more time then we'll merge that pr without it.

**Comment by szpak** [comment]
@ryangardner We want to merge #1075 this week. If you are unable to address the change requests made by Leonard by that time, we can understand that. Just let us know :).

**Comment by szpak** [comment]
@ryangardner I made required changes and it is no merged. Thanks for your PR.

**Comment by ryangardner** [comment]
oh, thanks! I just saw the most recent message but the notifications for this issue got swallowed up in my inbox for some reason

**Pull request merged**